### ISSUE TITLE
Cannot return iterators wrapping &mut ParserIterator

### ISSUE THREAD

**Issue by ambiso** [rust-bakery/nom#1586]
## Prerequisites

Here are a few things you should provide to help me understand the issue:

- Rust version : `rustc 1.67.0-nightly (53e4b9dd7 2022-12-04)`
- nom version : `7.1.1`
- nom compilation features used: `[]` or `["std"]`

## Test case

This code does not compile, but ideally should ([play](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f4c5a1f080a93926ac1f613fb87326bd)):

```rust
use nom::{
    character::complete::{digit1, newline},
    combinator::{iterator, opt},
    sequence::terminated,
    IResult,
};

fn parse_line(i: &str) -> IResult<&str, &str> {
    terminated(digit1, opt(newline))(i)
}

fn parse_input(i: &str) -> impl Iterator<Item = i32> + '_ {
    iterator(i, parse_line)
        // lots of processing, maybe a flat_map in between etc. - enough to warrant extracting it into a separate function
        .map(|x| x.parse::<i32>().unwrap())
}

fn main() {
    dbg!(parse_input("123\n456").collect::<Vec<_>>());
}
```

If I take a ParserIterator and use it as an Iterator, I can no longer return it from a function, since it's a `Map<&mut ParserIterator>`, and not a `Map<ParserIterator>`.

This is because the `Iterator` implementation of ParserIterator is on `&mut ParserIterator` and not `ParserIterator` directly.
See [here](https://github.com/Geal/nom/blob/324125f23ea7c1cca2c3677c4a2a454869f17127/src/combinator/mod.rs#L689).

My guess is that this was done because one might later want to call [`.finish()`](https://github.com/Geal/nom/blob/324125f23ea7c1cca2c3677c4a2a454869f17127/src/combinator/mod.rs#L680) on the parser iterator, and one can only do that if it hasn't been moved by a previous `Iterator::map(self, Fn)`.

However, if instead the Iterator implementation was direcly on `ParserIterator` instead of `&mut ParserIterator`, the above example would compile.
One could also decide not to move the iterator by calling [`.by_ref()`](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.by_ref) on the ParserIterator, before applying `.map()`.

Like this:

```rust
let it = iterator(i, parse_line);
it.by_ref()
  .map(|x| x.parse::<i32>().unwrap())
it.finish();
```

This would be as I understand it strictly more general, since it now allows for both returning a `Map<ParserIterator>`, and calling `.finish()` when needed (though not both at the same time), whereas the current variant excludes the possibility of returning a `Map<ParserIterator>`.

This proposal would however be a breaking change.

**Comment by Geal** [comment]
this makes sense. Since there will be a new major version of nom, it can come as a breaking change

**Comment by Geal** [comment]
fixed by #1587
